include/uapi/sound/sof-ipc.h

```diff
@@ -732,6 +732,18 @@ struct sof_ipc_comp_eq_iir {
 	struct sof_ipc_comp_config config;
 } __attribute__((packed));
 
+/** \brief Types of EFFECT */
+enum sof_ipc_effect_type {
+	SOF_EFFECT_INTEL_NONE = 0,	/**< None */
+	SOF_EFFECT_INTEL_EQFIR,		/**< Intel FIR */
+	SOF_EFFECT_INTEL_EQIIR,		/**< Intel IIR */
+};
+
+/* general purpose EFFECT configuration */
+struct sof_ipc_comp_effect {
+	enum sof_ipc_effect_type type;
+} __attribute__((packed));
```

Comment:
@lgirdwood @juimonen @singalsu 
Are you sure this definition is correct? All soc_ipc_comp_xxx structures start with struct sof_ipc_comp comp;
e.g. struct sof_ipc_comp_eq_fir {
	struct sof_ipc_comp comp;
	struct sof_ipc_comp_config config;
	uint32_t size;
	unsigned char data[0];
} __attribute__((packed));

here you are missing the comp field so all the address calculations will be off.
struct sof_ipc_comp_effect {
>>> missing comp field?
	enum sof_ipc_effect_type type;
} __attribute__((packed));

I wonder how this might work since there is no way the IPC can read the right fields, this will read uninitialized data.

While I am at it, is there really a need for this IPC? For example, for an EQ, you will have to rely on two IPC messages, one to set the effect type and one to set the actual parameters. Is this a desired feature and can the topology actually support this? I can't recall any other component where we do this.

Reply:
hi,
long time since I checked this... 

So as I understand the ipc_comp_effect is never sent to DSP it is mainly helper to used to parse the token from topology, so that we can then parse that particular effects data (which is sent to dsp). Like ipc_frame_type or ipc_dai_type. So actually I don't know why it is called "ipc". I think I copied the structure from the existing types.

And as I understood in the topology the ordering doesn't matter if you are just using tuples and not some byte fields etc?. So you can basically parse the tuples in any order you want...now I might be horribly wrong with this. I think I tested this to work with tuples and I remember that whole thing collapsed if you put bytes field to widget data. The effect type is actually after the general comp data in the topology.

As the dapm "effect" doesn't have any subcategories we need some way to differentiate between effects, like we have now already iir and fir. in the previous demo the dapm widget effect was hardcoded to be iir.